File: longest-substring-without-repeating-characters/njngwn.java
// Time Complexity: O(n), n: s.length()
// Space Complexity: O(n), n: s.length()
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int begin = 0;
        int end = 0;
        int maxLength = 0;

        Set<Character> letterSet = new HashSet<>();
        while (end < s.length()) {
            // if the letter is not in the hashset, then add it into hashset and move end pointer
            if (!letterSet.contains(s.charAt(end))) {
                letterSet.add(s.charAt(end));
                ++end;
                maxLength = Math.max(maxLength, end - begin); // update maxlength
            } else { // if the letter is in the hashset, then remove it and move begin pointer
                letterSet.remove(s.charAt(begin));
                ++begin;
            }
        }

        return maxLength;
    }
}

File: longest-consecutive-sequence/njngwn.py
class Solution:
    # Time Complexity: O(nlogn), n: len(nums)
    # Space Complexity: O(n), n: len(nums)
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums:
            return 0

        nums = sorted(set(nums)) # O(nlogn)
        longest = 1
        length = 1

        for i in range(1, len(nums)):
            if nums[i] - nums[i-1] == 1:
                length += 1
                longest = max(longest, length)
            else:
                length = 1

        return longest
